ignite/metrics/psnr.py

```diff
@@ -0,0 +1,104 @@
+from typing import Callable, Optional, Sequence, Union
+
+import torch
+import torch.nn.functional as F
+
+from ignite.exceptions import NotComputableError
+from ignite.metrics.metric import Metric, reinit__is_reduced, sync_all_reduce
+
+__all__ = ["PSNR"]
+
+
+class PSNR(Metric):
+    """Computes `Peak signal-to-noise ratio (PSNR) <https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio>`_.
+
+    Args:
+        data_range (float): Range of the image. If not provided, it will be determined
+            from the image type. Typically, ``1.0``.
+        output_transform (callable, optional): A callable that is used to transform the Engine’s
+            process_function’s output into the form expected by the metric.
+        device (str or torch.device): specifies which device updates are accumulated on.
+            Setting the metric’s device to be the same as your update arguments ensures
+            the update method is non-blocking. By default, CPU.
+
+    Example:
+
+    To use with ``Engine`` and ``process_function``, simply attach the metric instance to the engine.
+    The output of the engine's ``process_function`` needs to be in format of
+    ``(y_pred, y)`` or ``{'y_pred': y_pred, 'y': y, ...}``.
+
+    Note:
+        ``y_pred`` and ``y`` have to be normalized image tensors, should have the same shape / dtype,
+        should have BxCxHxW shape.
+
+    .. code-block:: python
+
+        def process_function(engine, batch):
+            # ...
+            return y_pred, y
+        engine = Engine(process_function)
+        psnr = PSNR()
+        psnr.attach(engine, "psnr")
+    """
+
+    def __init__(
+        self,
+        data_range: Optional[float] = None,
+        output_transform: Callable = lambda x: x,
+        device: Union[str, torch.device] = torch.device("cpu"),
+    ):
+        super().__init__(output_transform=output_transform, device=device)
+        self.data_range = data_range
+        self.dtype_range = {
+            torch.float16: (-1, 1),
+            torch.float32: (-1, 1),
+            torch.float64: (-1, 1),
+        }
```

Comment:
should we have `dtype_range` similar to
https://github.com/scikit-image/scikit-image/blob/master/skimage/util/dtype.py#L18-L33

Currently we only need `dtype_range` for `PSNR` and `SSIM` metrics AFAIK.

Reply:
Is this really useful considering that all min/max values are similar ?

In `slearn`, it makes sense because différent `dtype` are handled. Maybe I'm missing something?

Reply:
if the difference between the min and max values of the image (float type) is low (0.5/6/7), the final value is different from `sklearn` (starting from 2/3 decimal places.)

Reply:
I think @sdesrozis 's comment is about as the same data range: `(-1, 1)` for all floating types, we can directly put it without using dictionary structure: `dmin, dmax = -1.0, 1.0` (line 83). That, there is still a valid question if input dtype is not float ...